File: src/cache/favicons.ts
import { createHash } from 'node:crypto';
import http from 'node:http';
import https from 'node:https';
import { mkdir, readFile, writeFile } from 'node:fs/promises';
import path from 'node:path';
import { faviconDir } from './store.ts';

/** Icons are small; anything larger is not a favicon and we refuse it. */
const MAX_FAVICON_BYTES = 256 * 1024;
const FETCH_TIMEOUT = 3000;

const ALLOWED_TYPES = [
  'image/png',
  'image/x-icon',
  'image/vnd.microsoft.icon',
  'image/svg+xml',
  'image/jpeg',
  'image/gif',
  'image/webp',
  'image/avif',
];

export interface Favicon {
  hash: string;
  contentType: string;
  bytes: Buffer;
}

/** Sniff the format, because plenty of servers mislabel favicons. */
function detectType(bytes: Buffer, declared: string | undefined): string | undefined {
  if (bytes.length < 4) return undefined;

  if (bytes[0] === 0x89 && bytes[1] === 0x50 && bytes[2] === 0x4e && bytes[3] === 0x47) {
    return 'image/png';
  }
  if (bytes[0] === 0x00 && bytes[1] === 0x00 && bytes[2] === 0x01 && bytes[3] === 0x00) {
    return 'image/x-icon';
  }
  if (bytes[0] === 0xff && bytes[1] === 0xd8 && bytes[2] === 0xff) return 'image/jpeg';
  if (bytes.subarray(0, 3).toString('latin1') === 'GIF') return 'image/gif';
  if (
    bytes.subarray(0, 4).toString('latin1') === 'RIFF' &&
    bytes.subarray(8, 12).toString('latin1') === 'WEBP'
  ) {
    return 'image/webp';
  }

  // SVG is text, and may open with a comment, doctype or XML declaration.
  const head = bytes.subarray(0, 256).toString('utf8').trimStart();
  if (head.startsWith('<?xml') || head.startsWith('<svg') || head.includes('<svg')) {
    return 'image/svg+xml';
  }

  const base = declared?.split(';')[0]?.trim().toLowerCase();
  return base && ALLOWED_TYPES.includes(base) ? base : undefined;
}

/**
 * Fetch an icon over plain HTTP or TLS, ignoring certificate problems.
 *
 * This has to happen server-side. A browser asked to load an icon straight
 * from a self-signed HTTPS dev server refuses it, so the page would show a
 * broken image for exactly the servers this tool exists to catalogue.
 */
export function fetchFavicon(url: string, signal?: AbortSignal): Promise<Favicon | undefined> {
  return new Promise((resolve) => {
    let parsed: URL;
    try {
      parsed = new URL(url);
    } catch {
      return resolve(undefined);
    }

    const secure = parsed.protocol === 'https:';
    if (!secure && parsed.protocol !== 'http:') return resolve(undefined);

    const mod = secure ? https : http;
    const req = mod.request(
      url,
      {
        method: 'GET',
        ...(secure ? { rejectUnauthorized: false } : {}),
        headers: { accept: 'image/*,*/*;q=0.8', connection: 'close' },
        timeout: FETCH_TIMEOUT,
        ...(signal ? { signal } : {}),
      },
      (res) => {
        if (!res.statusCode || res.statusCode < 200 || res.statusCode >= 300) {
          res.resume();
          return resolve(undefined);
        }

        const chunks: Buffer[] = [];
        let size = 0;
        let aborted = false;

        res.on('data', (chunk: Buffer) => {
          size += chunk.length;
          if (size > MAX_FAVICON_BYTES) {
            aborted = true;
            res.destroy();
            return;
          }
          chunks.push(chunk);
        });

        const done = () => {
          if (aborted || chunks.length === 0) return resolve(undefined);
          const bytes = Buffer.concat(chunks);
          const contentType = detectType(bytes, res.headers['content-type']);
          // A 200 that is really an HTML error page is a very common way for
          // /favicon.ico to "succeed"; the sniff is what catches it.
          if (!contentType) return resolve(undefined);
          resolve({
            hash: createHash('sha256').update(bytes).digest('hex').slice(0, 16),
            contentType,
            bytes,
          });
        };

        res.on('end', done);
        res.on('close', done);
        res.on('error', () => resolve(undefined));
      },
    );

    req.on('timeout', () => req.destroy());
    req.on('error', () => resolve(undefined));
    req.end();
  });
}

function blobPath(hash: string): string {
  return path.join(faviconDir(), hash);
}

function metaPath(hash: string): string {
  return path.join(faviconDir(), `${hash}.type`);
}

/** Persist icon bytes so the UI can serve them from its own origin. */
export async function storeFavicon(icon: Favicon): Promise<void> {
  try {
    await mkdir(faviconDir(), { recursive: true });
    await Promise.all([
      writeFile(blobPath(icon.hash), icon.bytes),
      writeFile(metaPath(icon.hash), icon.contentType, 'utf8'),
    ]);
  } catch {
    // Losing a cached icon is cosmetic; never fail a scan over it.
  }
}

export async function readFavicon(
  hash: string,
): Promise<{ bytes: Buffer; contentType: string } | undefined> {
  // Hashes are hex from our own hashing, but this value reaches us from an
  // HTTP path later, so refuse anything that could escape the directory.
  if (!/^[0-9a-f]{16}$/.test(hash)) return undefined;
  try {
    const [bytes, contentType] = await Promise.all([
      readFile(blobPath(hash)),
      readFile(metaPath(hash), 'utf8').catch(() => 'application/octet-stream'),
    ]);
    return { bytes, contentType: contentType.trim() };
  } catch {
    return undefined;
  }
}

/** Fetch and cache in one step, returning the hash to store on the record. */
export async function cacheFavicon(
  url: string,
  signal?: AbortSignal,
): Promise<string | undefined> {
  const icon = await fetchFavicon(url, signal);
  if (!icon) return undefined;
  await storeFavicon(icon);
  return icon.hash;
}

File: src/cache/favicons.rs
//! Content-addressed favicon cache.
//!
//! Icons are fetched once and stored under their content hash, so the index
//! page can serve them from this machine rather than pointing the browser at a
//! dev server that may have stopped since.

use std::path::PathBuf;
use std::time::Duration;

use sha1::{Digest, Sha1};

use crate::config::cache_dir;

/// Icons are small. Anything larger is not an icon, whatever it claims.
const MAX_ICON_BYTES: usize = 256 * 1024;
const FETCH_TIMEOUT: Duration = Duration::from_millis(1500);

pub fn favicon_dir() -> PathBuf {
    cache_dir().join("favicons")
}

fn hash_bytes(bytes: &[u8]) -> String {
    let mut hasher = Sha1::new();
    hasher.update(bytes);
    let digest = hasher.finalize();
    let mut out = String::with_capacity(16);
    for byte in digest.iter().take(8) {
        out.push_str(&format!("{byte:02x}"));
    }
    out
}

/// A hash is a filename, so it must not be able to name anything else.
fn is_valid_hash(hash: &str) -> bool {
    hash.len() == 16 && hash.chars().all(|c| c.is_ascii_hexdigit())
}

/// Guess a content type from the bytes themselves.
///
/// The `Content-Type` a dev server puts on `/favicon.ico` is wrong often enough
/// that sniffing is more reliable, and the set of formats is tiny.
pub fn sniff_content_type(bytes: &[u8]) -> &'static str {
    if bytes.starts_with(b"\x89PNG\r\n\x1a\n") {
        "image/png"
    } else if bytes.starts_with(b"\x00\x00\x01\x00") {
        "image/x-icon"
    } else if bytes.starts_with(b"GIF8") {
        "image/gif"
    } else if bytes.starts_with(&[0xFF, 0xD8, 0xFF]) {
        "image/jpeg"
    } else if bytes.starts_with(b"RIFF") && bytes.get(8..12) == Some(b"WEBP") {
        "image/webp"
    } else {
        // SVG is text and can lead with a comment, a doctype or the tag.
        let head = String::from_utf8_lossy(&bytes[..bytes.len().min(256)]).to_lowercase();
        if head.contains("<svg") {
            "image/svg+xml"
        } else {
            "application/octet-stream"
        }
    }
}

/// Do the bytes look like an image at all?
///
/// A dev server with SPA routing answers `/favicon.ico` with its index page and
/// a 200, so without this the cache fills with copies of index.html.
fn looks_like_an_icon(bytes: &[u8]) -> bool {
    !bytes.is_empty() && sniff_content_type(bytes) != "application/octet-stream"
}

pub struct CachedIcon {
    pub bytes: Vec<u8>,
    pub content_type: &'static str,
}

pub fn read_favicon(hash: &str) -> Option<CachedIcon> {
    if !is_valid_hash(hash) {
        return None;
    }
    let bytes = std::fs::read(favicon_dir().join(hash)).ok()?;
    let content_type = sniff_content_type(&bytes);
    Some(CachedIcon {
        bytes,
        content_type,
    })
}

/// Fetch an icon and store it under its content hash.
///
/// Returns the hash, or None for anything that did not turn out to be an image.
pub async fn cache_favicon(url: &str) -> Option<String> {
    let bytes = fetch_bytes(url).await?;
    if !looks_like_an_icon(&bytes) {
        return None;
    }

    let hash = hash_bytes(&bytes);
    let dir = favicon_dir();
    let path = dir.join(&hash);

    // Content-addressed: identical bytes are already the same file.
    if path.exists() {
        return Some(hash);
    }
    std::fs::create_dir_all(&dir).ok()?;
    std::fs::write(&path, &bytes).ok()?;
    Some(hash)
}

/// GET a URL and return at most `MAX_ICON_BYTES` of body.
async fn fetch_bytes(url: &str) -> Option<Vec<u8>> {
    use http_body_util::BodyExt;

    let parsed: hyper::Uri = url.parse().ok()?;
    // Only loopback: a page can declare any favicon URL it likes, and fetching
    // an arbitrary remote one on its say-so is not our business.
    let host = parsed.host()?;
    let bare = host.trim_start_matches('[').trim_end_matches(']');
    let is_local = bare == "localhost"
        || bare.ends_with(".localhost")
        || bare
            .parse::<std::net::IpAddr>()
            .map(|ip| ip.is_loopback())
            .unwrap_or(false);
    if !is_local {
        return None;
    }
    // TLS to a dev server's self-signed cert is a whole handshake for an icon.
    if parsed.scheme_str() == Some("https") {
        return None;
    }

    let port = parsed.port_u16().unwrap_or(80);
    let stream = tokio::time::timeout(FETCH_TIMEOUT, tokio::net::TcpStream::connect((bare, port)))
        .await
        .ok()?
        .ok()?;

    let (mut sender, connection) =
        hyper::client::conn::http1::handshake(hyper_util::rt::TokioIo::new(stream))
            .await
            .ok()?;
    tokio::spawn(async move {
        let _ = connection.await;
    });

    let request = hyper::Request::builder()
        .uri(parsed.path_and_query().map(|p| p.as_str()).unwrap_or("/"))
        .header("host", format!("{host}:{port}"))
        .header("accept", "image/*")
        .header("connection", "close")
        .body(String::new())
        .ok()?;

    let response = tokio::time::timeout(FETCH_TIMEOUT, sender.send_request(request))
        .await
        .ok()?
        .ok()?;

    if !response.status().is_success() {
        return None;
    }

    let mut body = response.into_body();
    let mut collected = Vec::new();
    while let Some(frame) = body.frame().await {
        let Ok(frame) = frame else { break };
        if let Some(chunk) = frame.data_ref() {
            let room = MAX_ICON_BYTES.saturating_sub(collected.len());
            if room == 0 {
                break;
            }
            collected.extend_from_slice(&chunk[..chunk.len().min(room)]);
        }
    }

    (!collected.is_empty()).then_some(collected)
}

/// Delete cached icons nothing refers to any more.
pub fn prune(live: &std::collections::HashSet<String>) {
    let Ok(entries) = std::fs::read_dir(favicon_dir()) else {
        return;
    };
    for entry in entries.flatten() {
        let name = entry.file_name().to_string_lossy().into_owned();
        if !live.contains(&name) {
            let _ = std::fs::remove_file(entry.path());
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    const PNG: &[u8] = b"\x89PNG\r\n\x1a\n\x00\x00\x00\rIHDR";

    #[test]
    fn sniffs_the_formats_favicons_actually_use() {
        assert_eq!(sniff_content_type(PNG), "image/png");
        assert_eq!(
            sniff_content_type(b"\x00\x00\x01\x00\x01\x00"),
            "image/x-icon"
        );
        assert_eq!(sniff_content_type(b"GIF89a"), "image/gif");
        assert_eq!(sniff_content_type(&[0xFF, 0xD8, 0xFF, 0xE0]), "image/jpeg");
        assert_eq!(
            sniff_content_type(b"<?xml version=\"1.0\"?><svg xmlns=\"...\">"),
            "image/svg+xml"
        );
    }

    #[test]
    fn rejects_an_html_page_masquerading_as_an_icon() {
        // A dev server with SPA routing answers /favicon.ico with index.html
        // and a 200. Caching that would fill the cache with web pages.
        let html = b"<!doctype html><html><head><title>App</title></head></html>";
        assert!(!looks_like_an_icon(html));
        assert!(!looks_like_an_icon(b""));
        assert!(looks_like_an_icon(PNG));
    }

    #[test]
    fn hashes_are_content_addressed_and_stable() {
        let first = hash_bytes(PNG);
        assert_eq!(first, hash_bytes(PNG));
        assert_ne!(first, hash_bytes(b"different bytes"));
        assert!(is_valid_hash(&first), "{first} should be a valid hash");
    }

    #[test]
    fn rejects_hashes_that_could_name_another_file() {
        for hostile in [
            "../../etc/passwd",
            "not-hex-at-all!!",
            "",
            "abc",
            "0123456789abcdef0", // one too long
        ] {
            assert!(!is_valid_hash(hostile), "{hostile:?} should be refused");
            assert!(read_favicon(hostile).is_none());
        }
    }

    #[tokio::test]
    async fn refuses_to_fetch_icons_from_off_this_machine() {
        // A page declares its own favicon URL. Following one that points off
        // the box would make the scanner into someone else's HTTP client.
        assert!(fetch_bytes("http://example.com/favicon.ico")
            .await
            .is_none());
        assert!(fetch_bytes("http://8.8.8.8/favicon.ico").await.is_none());
    }

    #[tokio::test]
    async fn fetches_and_stores_a_real_icon() {
        use tokio::io::{AsyncReadExt, AsyncWriteExt};

        let temp = tempfile::tempdir().unwrap();
        std::env::set_var("XDG_CACHE_HOME", temp.path());

        let listener = tokio::net::TcpListener::bind("127.0.0.1:0").await.unwrap();
        let port = listener.local_addr().unwrap().port();
        tokio::spawn(async move {
            while let Ok((mut socket, _)) = listener.accept().await {
                tokio::spawn(async move {
                    let mut scratch = [0u8; 1024];
                    let _ = socket.read(&mut scratch).await;
                    let mut response = format!(
                        "HTTP/1.1 200 OK\r\ncontent-type: image/png\r\n\
                         content-length: {}\r\nconnection: close\r\n\r\n",
                        PNG.len()
                    )
                    .into_bytes();
                    response.extend_from_slice(PNG);
                    let _ = socket.write_all(&response).await;
                    let _ = socket.shutdown().await;
                });
            }
        });

        let hash = cache_favicon(&format!("http://127.0.0.1:{port}/favicon.png"))
            .await
            .expect("should cache the icon");

        let read_back = read_favicon(&hash).expect("should read it back");
        assert_eq!(read_back.bytes, PNG);
        assert_eq!(read_back.content_type, "image/png");

        std::env::remove_var("XDG_CACHE_HOME");
    }
}
